File: M/2293-min-max-game/main.go
package main

func minMaxGame(nums []int) int {
	cnt := len(nums)
	buf := make([]int, len(nums))
	copy(buf, nums)
	for len(buf) > 1 {
		t := make([]int, cnt/2)
		s := 0
		for i := 0; i < len(buf); i = i + 2 {
			if s%2 == 0 {
				v := buf[i]
				if v > buf[i+1] {
					v = buf[i+1]
				}
				t[i/2] = v
			} else {
				v := buf[i]
				if v < buf[i+1] {
					v = buf[i+1]
				}
				t[i/2] = v
			}
			s++
		}
		buf = t
	}
	return buf[0]
}

File: M/2293-min-max-game/main.rs
impl Solution {
  pub fn min_max_game(nums: Vec<i32>) -> i32 {
    let mut n = nums;
    let mut size: usize = n.len() / 2;

    while size >= 1 {
      (0..size).for_each(|idx| {
        n[idx] = match idx % 2 {
          0 => n[2 * idx].min(n[2 * idx + 1]),
          1 => n[2 * idx].max(n[2 * idx + 1]),
          _ => 0,
        }
      });
      size /= 2
    }
    n[0]
  }
}
